src/charm.py

```diff
@@ -470,6 +476,14 @@ def reconcile(self, event):
             self.generate_tokens()
             self.configure_observability()
             self.apply_node_labels()
+            self.open_ports()
+
+    def open_ports(self):
+        """Open control plane ports needed for remote access to the cluster."""
+        try:
+            self.unit.open_port("tcp", self.APISERVER_PORT)
+        except ops.ModelError:
+            log.warning(f"Failed to open port {self.APISERVER_PORT}. Will retry.")
```

Comment:
the reconciler would rather you do:

```suggestion
    @status.on_error(ops.WaitingStatus("Waiting to open port"))
    def open_ports(self):
        """Open control plane ports needed for remote access to the cluster."""
        self.unit.open_port("tcp", self.APISERVER_PORT)
```

Reply:
this decorator will catch that exception, log it, and hold off on "reconciling" the charm because of the failure.  This will mean it ACTUALLY is retried.   What's missing from this suggestion is your specific text around "Failed to open port" -- though a similar message and stack trace will be in your debug log and the charm will go into a `WaitingStatus`

Reply:
Splendid!  thanks for the suggestion.